src/pluginextension.c

```diff
@@ -0,0 +1,136 @@
+/*
+ *      pluginextension.c - this file is part of Geany, a fast and lightweight IDE
+ *
+ *      Copyright 2023 The Geany contributors
+ *
+ *      This program is free software; you can redistribute it and/or modify
+ *      it under the terms of the GNU General Public License as published by
+ *      the Free Software Foundation; either version 2 of the License, or
+ *      (at your option) any later version.
+ *
+ *      This program is distributed in the hope that it will be useful,
+ *      but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *      GNU General Public License for more details.
+ *
+ *      You should have received a copy of the GNU General Public License along
+ *      with this program; if not, write to the Free Software Foundation, Inc.,
+ *      51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "pluginextension.h"
+
+
+static gboolean func_return_false(GeanyDocument *doc)
+{
+	return FALSE;
+}
+
+
+static GPtrArray *func_return_ptrarr(GeanyDocument *doc)
+{
+	return NULL;
+}
+
+
+static void func_args_doc_bool(GeanyDocument *doc, gboolean dummy1)
+{
+}
+
+
+static void func_args_doc_int_bool(GeanyDocument *doc, gint dummy1, gboolean dummy2)
+{
+}
+
+
+static PluginExtension dummy_extension = {
+	.autocomplete_provided = func_return_false,
+	.autocomplete_perform = func_args_doc_bool,
+
+	.calltips_provided = func_return_false,
+	.calltips_show = func_args_doc_bool,
+
+	.goto_provided = func_return_false,
+	.goto_perform = func_args_doc_int_bool,
+
+	.doc_symbols_provided = func_return_false,
+	.doc_symbols_get = func_return_ptrarr,
+
+	.symbol_highlight_provided = func_return_false
+};
+
+static PluginExtension *current_extension = &dummy_extension;
+
+
+GEANY_API_SYMBOL
+void plugin_extension_register(PluginExtension *extension)
+{
+	/* possibly, in the future if there's a need for multiple extensions,
+	 * have a list of extensions and add/remove to/from the list */
+	current_extension = extension;
```

Comment:
Maybe add a warning if the active extension isn't the default one, so 2 conflicting plugins don't go entirely unnoticed?

Reply:
Or really just add the list here. What I think in principle should happen is something like this (demonstrated on goto definition/declaration):
```C
gboolean plugin_extension_goto_provided(GeanyDocument *doc)
{
	PluginExtension *extension;
	gint i;

	foreach_ptr_array(extension, i, all_extensions)
	{
		if (extension->goto_provided && extension->goto_provided(doc))
			return TRUE;
	}
	return FALSE;
}


void plugin_extension_goto_perform(GeanyDocument *doc, gint pos, gboolean definition)
{
	PluginExtension *extension;
	gint i;

	foreach_ptr_array(extension, i, all_extensions)
	{
		if (extension->goto_provided && extension->goto_provided(doc))
		{
			extension->goto_perform(doc, pos, definition);
			return;
		}
	}
}
```

Basically the first item in the list providing the functionality for the document filetype would "win". The `perform()` of other plugins won't be called. This might require some cooperation among plugins so they allow disabling certain functionality and allow the other plugins do what they want. This is possible with the LSP plugin already where one can set `goto_enable=false` so `plugin_extension_goto_provided()` would return `FALSE` which would allow the other plugin in the list provide the goto feature.

Reply:
Just a note that if we want to use this interface for Geany itself, Geany's PluginExtension could be the last one in the chain and serve as a fallback.